drv/lpc55-spi-server/src/main.rs

```diff
@@ -63,74 +369,339 @@ fn main() -> ! {
+                    || rctx.state == RxState::Payload
+                {
+                    // XXX include a more useful error response?
+                    ringbuf_entry!(Trace::RxFragment);
+                    if tctx.enqueue(MsgType::ErrorRsp, EnqueueBuf::Empty) {
+                        ringbuf_entry!(Trace::CannotSendRxFragError);
+                        rctx.state = RxState::Dispatch;
+                    }
+                }
+                // Process transitory states
+                if rctx.state == RxState::Dispatch {
+                    // Receive is complete.
+                    // Processing may take a long time and detract from keeping up
+                    // with Tx and Rx FIFOs.
+                    // Processing has been delayed until CSn was de-asserted.
+                    rctx.state = RxState::Responding;
+                    ringbuf_entry!(Trace::Responding);
+
+                    if tctx.state == TxState::Idle {
+                        let rmsg = Msg::parse(&mut *rctx.rx).unwrap_lite();
+
+                        // Errors and SP always being allowed to send a command
+                        // on CSn assert means that there can be corner cases.
+                        // XXX Is the Tx FIFO empty or is the result of a previous
+                        // command still queued?
+
+                        // Invalid message protocol and length are already
+                        // handled in rx_byte().
+                        match rmsg.msgtype() {
+                            MsgType::StatusReq => {
+                                ringbuf_entry!(Trace::RespondToStatus);
+                                tctx.enqueue(
+                                    MsgType::StatusRsp,
+                                    EnqueueBuf::Copy(&bootrom_crc32[..]),
+                                );
+                                again = true;
+                            }
+                            MsgType::EchoReq => {
+                                ringbuf_entry!(Trace::RespondToEcho);
+                                tctx.enqueue(
+                                    MsgType::EchoRsp,
+                                    EnqueueBuf::Copy(
+                                        rmsg.payload_get().unwrap_lite(),
+                                    ),
+                                );
+                                again = true;
+                            }
+                            _ => {
+                                // The message received was an unknown type.
+                                ringbuf_entry!(Trace::RespondToUnknown);
+                                tctx.enqueue(
+                                    MsgType::ErrorRsp,
+                                    EnqueueBuf::Empty,
+                                );
+                            }
+                        }
+                    }
+                }
+            }
             if !again {
```

Comment:
It seems like the inner loop can run multiple times without returning to the kernel, and the inner loop also checks and clears interrupt flags on its own.  However, the inner loop leaves (hardware) interrupts on, which means we could have a somewhat confusing series of events:

- Inner loop runs, setting `again = true`
- Interrupt fires (e.g. SSA)
  - Kernel marks that interrupt as fired
- Inner loop runs again, noticing the interrupt and clearing its bit
- Inner loop ends
- Return to outer loop, re-enabling irqs from the kernel
- The kernel immediately notifies us of the IRQ that we already saw + handled

This suggests to me that we should keep (hardware) interrupts disabled when in the inner loop, then re-enable them before enabling (kernel) interrupts.